draft-ietf-tls-tls13.md

```diff
@@ -3025,17 +2937,263 @@ processing the server's Certificate, CertificateVerify, and
+                       MUST allow the key to be used for signing with the
+                       allow the key to be used for signing with the
+                       hash algorithm that will be employed in the
+                       server's KeyShare extension.  The public key
+                       MUST use a curve and point format supported by
+                       the client, as described in [RFC4492].
+~~~~
+
+- The "server_name" and "trusted_ca_keys" extensions {{RFC6066}} are used to
+  guide certificate selection. As servers MAY require the presence of the server_name
+  extension, clients SHOULD send this extension.
+
+All certificates provided by the server MUST be signed by a
+hash/signature algorithm pair that appears in the "signature_algorithms"
+extension provided by the client, if they are able to provide such
+a chain (see {{signature-algorithms}}).
+If the server cannot produce a certificate chain that is signed only via the
+indicated supported pairs, then it SHOULD continue the handshake by sending
+the client a certificate chain of its choice that may include algorithms
+that are not known to be supported by the client. This fallback chain MAY
+use the deprecated SHA-1 hash algorithm.
+If the client cannot construct an acceptable chain using the provided
+certificates and decides to abort the handshake, then it MUST send an
+"unsupported_certificate" alert message and close the connection.
+
+If the server has multiple certificates, it chooses one of them based on the
+above-mentioned criteria (in addition to other criteria, such as transport
+layer endpoint, local configuration and preferences). If the server has a
+single certificate, it SHOULD attempt to validate that it meets these criteria.
+
+As cipher suites that specify new key exchange methods are specified for the
+TLS protocol, they will imply the certificate format and the required encoded
+keying information.
+
+
+##### Client Certificate Selection
+
+The following rules apply to certificates sent by the client:
+
+In particular:
+
+- The certificate type MUST be X.509v3 {{RFC5280}}, unless explicitly negotiated
+  otherwise (e.g., {{RFC5081}}).
+
+- If the certificate_authorities list in the certificate request
+  message was non-empty, one of the certificates in the certificate
+  chain SHOULD be issued by one of the listed CAs.
+
+- The certificates MUST be signed using an acceptable hash/
+  signature algorithm pair, as described in {{certificate-request}}.  Note
+  that this relaxes the constraints on certificate-signing
+  algorithms found in prior versions of TLS.
+
+- If the certificate_extensions list in the certificate request message
+  was non-empty, the end-entity certificate MUST match the extension OIDs 
+  recognized by the client, as described in {{certificate-request}}.
+
+Note that, as with the server certificate, there are certificates that use
+algorithms/algorithm combinations that cannot be currently used with TLS.
```

Comment:
another pre-existing oddity: "algorithms/algorithm" ? Shouldn't this just be "algorithm"?